src/OrleansRuntime/GrainTypeManager/GrainTypeData.cs

```diff
@@ -151,5 +154,11 @@ internal static PlacementStrategy GetPlacementStrategy(Type grainClass, Placemen
 
             return predicate;
         }
+
+        private ParameterInfo[] GetConstructorParameters()
+        {
+            var constructor = this.Type.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault();
+            return constructor?.GetParameters() ?? EmptyParameters;
+        }
```

Comment:
>IServiceProvider will pick a specific constructor, but only if the registration just identified the type. But of course you can also even provide a factory method, so you don't even need an accessible constructor at all

We're adding features that we intend to expose via constructor injection, so if one is to follow the pattern we're setting up, well defined constructor selection is necessary.  The standards we've adopted for our supported DI containers, also have a well defined constructor selection method.  So where is the problem here?

> I'm 100% it is the case that ASP.NET doesn't put that constraint on which constructor to use from a controller. 

Conforming to ASP.NET's controller pattern is not a goal.  The feature set we wish to support is the goal.  If we can achieve this using ASP.NET's controller pattern, that would be great as it is more widely understood, but we're dealing with asynchronously constructed grains in a distributed system, there not the same as controllers.